overlays/nodejs/s3/s3Mixins.ts

```diff
@@ -0,0 +1,232 @@
+    };
+    responseElements: {
+        "x-amz-request-id": string;
+        "x-amz-id-2": string;
+    };
+    s3: {
+        s3SchemaVersion: string;
+        configurationId: string;
+        bucket: {
+            name: string;
+            ownerIdentity: {
+                principalId: string;
+            },
+            arn: string;
+        };
+        object: {
+            key: string;
+            size: number;
+            eTag: string;
+            versionId?: string;
+            sequencer: string;
+        };
+    };
+}
+
+export type BucketEventHandler = lambda.EventHandler<BucketEvent, void>;
+
+interface SubscriptionInfo {
+    name: string;
+    events: string[];
+    filterPrefix?: string;
+    filterSuffix?: string;
+    lambdaFunctionArn: pulumi.Output<string>;
+    permission: lambda.Permission;
+}
+
+let bucketSubscriptionInfos = new Map<Bucket, SubscriptionInfo[]>();
+
+/**
+ * A component corresponding to a single underlying aws.s3.BucketNotification created for a bucket.
+ * Note: due to the AWS requirement that all notifications for a bucket be defined at once, the
+ * actual aws.s3.BucketNotification instances will only be created once the pulumi program runs to
+ * completion and all subscriptions have been heard about.
+ */
+export class BucketEventSubscription extends lambda.EventSubscription {
+    public readonly bucket: pulumi.Output<Bucket>;
+
+    public constructor(
+        name: string, bucket: Bucket, handler: BucketEventHandler,
+        args: BucketEventSubscriptionArgs, opts?: pulumi.ResourceOptions) {
+
+        super("aws:s3:BucketEventSubscription", name, { bucket: bucket }, opts);
+
+        this.func = lambda.createFunctionFromEventHandler(name, handler, { parent: this });
+
+        const permission = new lambda.Permission(name, {
+            function: this.func,
+            action: "lambda:InvokeFunction",
+            principal: "s3.amazonaws.com",
+            sourceArn: bucket.id.apply(bucketName => `arn:aws:s3:::${bucketName}`),
+        }, { parent: this });
+
+        this.permission = permission;
+
+        // We must create only a single BucketNotification per Bucket per AWS API limitations.  See
+        // https://github.com/terraform-providers/terraform-provider-aws/issues/1715.  So we push
+        // the subscription information here, and then actually create the BucketNotification if
+        // needed on process `beforeExit`.
+        let subscriptions = bucketSubscriptionInfos.get(bucket);
+        if (!subscriptions) {
+            subscriptions = [];
+            bucketSubscriptionInfos.set(bucket, subscriptions);
+        }
+
+        subscriptions.push({
+            name: name,
+            events: args.events,
+            filterPrefix: args.filterPrefix,
+            filterSuffix: args.filterSuffix,
+            lambdaFunctionArn: this.func.arn,
+            permission: permission,
+        });
+    }
+}
+
+process.on("beforeExit", () => {
```

Comment:
Is this still necessary, you think? Do you know why this has to be run before exit?

Reply:
Yeah, this is necessary.  The main issue here is that AWS requires all bucket subscriptions to be set in one swell foop.  So we have to collect them all over the life of the app, and then only apply right at the end.

Reply:
Specifically - a Bucket can only have one BucketNotification resource attached to it (which can itself specify multiple handler).  Since we want to allow these to be provided at mutiple points in time via a more imprative API, we have to collect these and create the BucketNotification at shutdown.

Arguably, this is one of the pieces of evidence suggesting we should be generally delaying more things, to allow other imperative models - but that's something for another day.